![Universidade Federal Rural do Rio de Janeiro](https://portal.ufrrj.br/wp-content/themes/portalufrrj/images/logomarca_ufrrj_cor.png)

---

### Programação Linear 2023-2
#### *Nova Iguaçu, 11 de novembro de 2023*
#### **Professor:** Erito Marques
#### **Alunos:** 
&emsp;Raphael Balmant

&emsp;Elis Braga

---

# Introdução

O presente relatório descreve a aplicação do método _Linear Programming Discriminant Analysis (LPDA)_ conforme proposto no artigo intitulado **"LPDA: A new classification method based on linear programming"**. O objetivo principal deste trabalho é desenvolver um modelo de classificação binária utilizando o LPDA em um conjunto de dados previamente selecionado, disponível na pasta "data".

O artigo também apresenta uma biblioteca em R que oferece métodos para a implementação do LPDA. A documentação completa dessa biblioteca pode ser encontrada no CRAN, no seguinte endereço: [pacote lpda](https://cran.r-project.org/web/packages/lpda/index.html).

No entanto, devido às complexidades associadas à conversão do código R para Python e às particularidades das estruturas de dados utilizadas pela biblioteca, optou-se por empregar a biblioteca [Pulp](https://coin-or.github.io/pulp/). Essa biblioteca possibilita a descrição iterativa do problema de programação linear a ser resolvido, oferecendo a opção de utilizar o solver de preferência do usuário. Outras ferramentas, como [AMPL](https://ampl.com/), seguem uma abordagem semelhante.

O método em questão foi empregado no contexto do diagnóstico de doença cardíaca, utilizando um dataset proveniente do Heart Disease Database, disponibilizado no Kaggle.

Este conjunto de dados compreende um total de 76 atributos, embora investigações prévias tenham se centrado em um subconjunto específico contendo 14 deles. Notadamente, até o momento presente, a preferência dos pesquisadores em aprendizado de máquina recai exclusivamente sobre o banco de dados originário de Cleveland.

A coluna "target" assume o papel de nossa variável de interesse, associada à presença de doença cardíaca no paciente, expressa por valores booleanos: 1 indicando a presença da doença e 0, a sua ausência.

O conjunto de dados em análise apresenta uma série de variáveis relevantes para o diagnóstico de doença cardíaca, cada uma caracterizando aspectos específicos relacionados à condição cardíaca dos indivíduos estudados. Abaixo, são detalhadas as informações de algumas dessas variáveis:

1. **Idade (AGE):** Esta variável, expressa em anos, varia no intervalo de 29 a 77 anos, fornecendo uma visão da distribuição etária dos pacientes sob investigação.

2. **Sexo (SEX):** Representada de maneira binária, onde o valor 1 indica o sexo masculino e o valor 0, o feminino, esta variável possibilita a análise de possíveis disparidades entre os gêneros na ocorrência de doença cardíaca.

3. **Tipo de Dor no Peito (CP):** Com quatro categorias distintas, este atributo fornece informações sobre a natureza da dor torácica relatada pelos pacientes:
    - Valor 1: Angina típica.
    - Valor 2: Angina atípica.
    - Valor 3: Sem dor anginal.
    - Valor 4: Assintomático.

4. **Pressão Arterial em Repouso (TRESTBPS):** Expressa em milímetros de mercúrio (mm Hg), esta medida oferece insights sobre a pressão arterial dos indivíduos em estado de repouso.

5. **Colesterol no Soro Sanguíneo (CHOL):** Medido em miligramas por decilitro (mg/dl), este atributo fornece informações sobre os níveis de colesterol no sangue dos pacientes.

6. **Concentração de Açúcar no Sangue em Jejum (FBS):** Representado como um valor booleano, onde 1 indica uma concentração de açúcar no sangue superior a 120 mg/dl e 0, inferior, este atributo avalia a presença de glicose elevada.

7. **Resultado da Eletrocardiografia em Repouso (RESTECG):** Classificado em três categorias distintas, este atributo proporciona uma avaliação do estado do coração em repouso:
    - Valor 0: Normal.
    - Valor 1: Com onda ST-T anormal.
    - Valor 2: Mostrando provável (ou definida) hipertrofia do ventrículo esquerdo.

8. **Máxima Taxa de Batimento Cardíaco Atiginda (THALACH):** Este atributo indica a taxa máxima de batimento cardíaco atingida durante o exercício.

9. **Angina Induzida por Exercício (EXANG):** Representado como um valor booleano, onde 1 indica a presença de angina induzida por exercício e 0, a sua ausência.

10. **Depressão ST Induzida por Exercício Relativamente Sossegado (OLDPEAK):** Este atributo mede a depressão do segmento ST induzida por exercício em relação ao repouso.

11. **Inclinação da Extremidade do Segmento ST no Exercício (SLOPE):** Classificado em três categorias, este atributo indica a inclinação da extremidade do segmento ST durante o exercício:
    - Valor 1: Inclinado para cima.
    - Valor 2: Plano.
    - Valor 3: Inclinado para baixo.

12. **Número de Vasos Coloridos pela Fluoroscopia (CA):** Este atributo representa o número de vasos sanguíneos coloridos pela fluoroscopia, variando de 0 a 3.

13. **Talassemias (THAL):** Classificado em três categorias, este atributo indica a condição talassêmica:
    - Valor 3: Normal.
    - Valor 6: Defeito fixo (irreparável).
    - Valor 7: Defeito reversível (reparável).

# Metodologia

A implementação do LPDA utilizando a biblioteca Pulp permitiu resolver o problema de programação linear, encontrando a solução ótima. Os coeficientes do plano H descritos no artigo, bem como a constante 'b', foram obtidos como resultado. Adicionalmente, foram desenvolvidos métodos para armazenar esses valores em disco, possibilitando sua reutilização posterior.

Com o arquivo do modelo armazenado, foi possível criar um servidor de aplicação que oferece duas rotas principais:

1. **Treinamento:** A rota `POST /heart/train` realiza o treinamento do modelo. Exemplo de chamada:
   ```bash
   !curl --request POST --url http://localhost:8080/heart/train
2. **Predição:** A rota `POST /heart/predict` permite realizar previsões com base no modelo treinado. Exemplo de chamada:

   ```bash
   !curl --request POST \
   --url http://localhost:8080/heart/predict \
   --header 'Content-Type: application/json' \
   --data '{
    "age": 19,
    "sex": 0,
    "cp": 1,
    "trestbps": 120,
    "chol": 204,
    "fbs": 0,
    "restecg": 0,
    "thalach": 172,
    "exang": 0,
    "oldpeak": 1.4,
    "slope": 2,
    "ca": 0,
    "thal": 2
   }'

In [1]:
import pandas as pd 
from train.lpda import lpda, simple_predict, predict
from model import model

In [2]:
file = '..\data\heart_disease\heart.csv'
df = pd.read_csv(file)
df.head()

<>:1: SyntaxWarning: invalid escape sequence '\d'
<>:1: SyntaxWarning: invalid escape sequence '\d'
C:\Users\elisb\AppData\Local\Temp\ipykernel_10344\1262763081.py:1: SyntaxWarning: invalid escape sequence '\d'
  file = '..\data\heart_disease\heart.csv'


age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


# Resultados

A partir da aplicação das funções definidas, foi possível aplicar o modelo para o dataset selecionado

In [4]:
result = model(df, 'target')

In [5]:
solution = result.solution
test = result.test
scores = result.scores

a = solution['a']
b = solution['b']

X_test = test['X_test']
Y_test = test['y_test']

precision = scores['precision']
accuracy = scores['accuracy']
f1 = scores['f1']

Conseguimos dessa forma obter as soluções ótimas para o problema dado por a e b:

In [6]:
print(f'Vetor de soluções a: {a}\n')
print(f'solução b: {b}\n')

Vetor de soluções a: [0.013748104, 1.1251695, 0.0, 0.0, 0.0039203577, 0.0, 0.0, 0.0, 1.3843964, 0.46359572, 0.0, 0.70430837, 0.45532538]

solução b: 4.9093753



Conjuntos de Teste para Avaliação do Modelo

Os conjuntos **X_test** e **y_test** desempenham um papel crucial na avaliação do desempenho do modelo proposto. Esses conjuntos foram derivados do conjunto de dados original e são estrategicamente utilizados para testar a capacidade do modelo de generalização para dados não observados.

**Matriz de Características (X_test):** O conjunto X_test consiste em uma matriz de características, em que cada linha representa uma entrada do conjunto de teste. Cada coluna dessa matriz corresponde a um atributo específico associado a uma instância de dados. Por exemplo, a primeira linha de X_test contém valores que representam atributos como idade, gênero, pressão arterial e outros, proporcionando um conjunto abrangente de informações para cada ponto de dados.

**Rótulos de Classe (y_test):** O conjunto y_test é uma matriz que contém os rótulos de classe associados a cada entrada em X_test. Cada valor em y_test representa a classe à qual a instância correspondente em X_test pertence. Este é um componente fundamental para avaliar a capacidade do modelo em categorizar corretamente as instâncias do conjunto de teste.

**Importância na Avaliação do Modelo:** Esses conjuntos são exclusivamente utilizados no cálculo de métricas de desempenho do modelo, tais como precisão, acurácia e pontuação F1. Ao analisar o desempenho do modelo em dados não vistos, podemos obter insights valiosos sobre sua capacidade de generalização. É importante destacar que a avaliação do modelo em um conjunto de teste independente ajuda a evitar problemas de sobreajuste e fornece uma visão mais realista de seu comportamento em condições do mundo real.

In [9]:
print(f'x_teste: {X_test}\n')

x_teste: [[5.90e+01 1.00e+00 3.00e+00 1.60e+02 2.73e+02 0.00e+00 0.00e+00 1.25e+02
  0.00e+00 0.00e+00 2.00e+00 0.00e+00 2.00e+00]
 [4.60e+01 1.00e+00 0.00e+00 1.20e+02 2.49e+02 0.00e+00 0.00e+00 1.44e+02
  0.00e+00 8.00e-01 2.00e+00 0.00e+00 3.00e+00]
 [6.40e+01 1.00e+00 2.00e+00 1.40e+02 3.35e+02 0.00e+00 1.00e+00 1.58e+02
  0.00e+00 0.00e+00 2.00e+00 0.00e+00 2.00e+00]
 [5.80e+01 1.00e+00 0.00e+00 1.28e+02 2.16e+02 0.00e+00 0.00e+00 1.31e+02
  1.00e+00 2.20e+00 1.00e+00 3.00e+00 3.00e+00]
 [6.40e+01 1.00e+00 2.00e+00 1.25e+02 3.09e+02 0.00e+00 1.00e+00 1.31e+02
  1.00e+00 1.80e+00 1.00e+00 0.00e+00 3.00e+00]
 [5.50e+01 1.00e+00 0.00e+00 1.60e+02 2.89e+02 0.00e+00 0.00e+00 1.45e+02
  1.00e+00 8.00e-01 1.00e+00 1.00e+00 3.00e+00]
 [4.70e+01 1.00e+00 2.00e+00 1.08e+02 2.43e+02 0.00e+00 1.00e+00 1.52e+02
  0.00e+00 0.00e+00 2.00e+00 0.00e+00 2.00e+00]
 [5.90e+01 1.00e+00 2.00e+00 1.26e+02 2.18e+02 1.00e+00 1.00e+00 1.34e+02
  0.00e+00 2.20e+00 1.00e+00 1.00e+00 1.00e+00]
 [5.80e+01 1.00

In [10]:
print(f'y_teste: {Y_test}')

y_teste: [[0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]


Abaixo, são apresentados os resultados das métricas empregadas na análise de precisão do modelo:

In [8]:
print(f'Precision: {(precision * 100):.2f}%\nAccuracy: {(accuracy * 100):.2f}%\nF1 Score: {(f1 * 100):.2f}% ')

Precision: 82.76%
Accuracy: 77.05%
F1 Score: 77.42% 


# Conclusão

O método revelou eficácia notável durante a análise de um conjunto de dados de menor escala, composto por 303 entradas, onde sua performance excedeu a marca de 70%. Entretanto, ao transpor esse método para um conjunto de dados mais amplo e robusto, observou-se uma significativa queda no desempenho. Este declínio sugere que, embora o método demonstre promissor em contextos mais limitados, sua capacidade de generalização e eficácia podem ser desafiadas quando confrontado com datasets mais extensos. Essa observação destaca a importância de considerar a escalabilidade e a adaptabilidade do método em diferentes cenários de aplicação.

# Referências

1. Nueda, M. J., Gandía, C., Molina, M. D. (2022). *LPDA: A new classification method based on linear programming.* PLOS ONE, 17(7), 1-13. DOI: [10.1371/journal.pone.0270403](https://doi.org/10.1371/journal.pone.0270403)  
    [LPDA on GitHub](https://github.com/mjnueda/lpda)

2. [Maxwell Document](https://www.maxwell.vrac.puc-rio.br/9758/9758_3.PDF)

3. Kaggle. (n.d.). *Heart Disease Dataset on Kaggle.*  
   [Kaggle Dataset](https://www.kaggle.com/datasets/yasserh/heart-disease-dataset/data)

4. Documentação:
   - Pandas. (n.d.). *Pandas Documentation.*  
     [Pandas Documentation](https://pandas.pydata.org/docs/)

   - Scikit-learn. (n.d.). *Scikit-learn Documentation.*  
     [Scikit-learn Documentation](https://scikit-learn.org/stable/)

   - NumPy. (n.d.). *NumPy Documentation.*  
     [NumPy Documentation](https://numpy.org/doc/)

   - Pulp. (n.d.). *Pulp Documentation.*  
     [Pulp Documentation](https://coin-or.github.io/pulp/CaseStudies/index.html)

   - FastAPI. (n.d.). *FastAPI Documentation.*  
     [FastAPI Documentation](https://fastapi.tiangolo.com/)

